### ISSUE TITLE
Allow shorthand for directives like `lwc:ref`

### ISSUE THREAD

**Issue by AllanOricil** [salesforce/lwc#3303]
**Is your feature request related to a problem? Please describe.**

no, I just would like to know if it could be simplified. And nice feature, by the way. We can get rid of `data-id` parameters in our templates.

https://developer.salesforce.com/blogs/2023/01/lwc-enhancements-for-developers-learn-moar-spring-23
![image](https://user-images.githubusercontent.com/55927613/213700483-360a018b-8bf4-4356-8f4e-005263d06675.png)

Is there a reason for prepending "ref" with the word "lwc:" ? 

Couldn't it just be

````html
<div :ref="reference"></div>
````
````html
<div ref="reference"></div>
````

Vue and React template engines are able to use just "ref", so I believe lwc would also be able to do it. It would be good if you could try to follow what other frameworks did well, so that people that are used to one of them can learn lwc faster. I only see benefits doing so.

Another question, does it accept expressions? 

Like,

````html
<div ref={dynamicRef}></div>
````
````html
<template for:each={bla} for:item="blas">
    <div ref={bla}></div>
</template>
````
````html
<template for:each={bla} for:item="blas">
     <div ref={getDynamicRef(bla)}></div>
</template>
````

**Comment by nolanlawson** [comment]
Thanks for the feedback!

> Is there a reason for prepending "ref" with the word "lwc:" ?

Yep, as mentioned in https://github.com/salesforce/lwc/issues/2503#issuecomment-1398793158, this is to make it clear what is native to the framework and what is part of LWC. This tells developers whether they should be searching for documentation on MDN or on Salesforce/lwc.dev. 🙂 

> Another question, does it accept expressions?

No. `lwc:ref` must be a static string value. The `for:each` use case is something [we specifically called out in the RFC](https://rfcs.lwc.dev/rfcs/lwc/0122-refs#loops) as not yet implemented, but worth implementing. I opened an issue to track that: https://github.com/salesforce/lwc/issues/3304

Closing this issue, as we don't currently plan to shorten `lwc:ref` to `ref`. Thanks for the feedback!

**Comment by nolanlawson** [comment]
Reopening this. After some internal discussion, it seems that it may be reasonable to support `:ref` in addition to `lwc:ref`. In fact, this is doable for all the `lwc:*` properties: `:if`, `:render-mode`, `:preserve-comments`, etc.

This might also be a good time to harmonize some of the directives that don't start with `lwc:`, e.g.:

- `for:each` -> `lwc:each` (/ht @jmsjtu)
- `key` -> `lwc:key`

The above two could be done in a backwards-compatible way; they would just be synonyms.

**Comment by AllanOricil** [comment]
`for:each` becoming `lwc:each` feels not natural. 

In my opinion, reading `for:each={contacts}` has always been weird because when I say the word `each` in english, the next word must be a singular one. For instance, I can't write "for each bananas". This directive expects a data type that represents a collection. And a collection is always named using a plural word.
With that said, the argument is that you should remove the word `each` because it does not feel grammarly correct or natural.

I would love if you could actually change the for:each directive entirely. Something like `for:each={(item, index) in collection}` would be much better because it would reduce the amount of characters in a template, while also respecting the English grammar and keeping the word `each`.

But if you can't really change it. What about using `lwc:for` and `lwc:item`? Then the correspondent shorthands would be:
`:for` and `:item`?

**Comment by caridy** [comment]
I'm up for shorting `lwc:` to just `:`, but I will probably continue to use `lwc:` myself, so it is very clear that this is intended to be LWC specific... the namespacing of HTML attributes is very clear, we namespace to prevent collisions.

**Comment by pmdartus** [comment]
As always thanks for raising this @AllanOricil. 

I am also up for finishing the migration of all the LWC directives to `lwc:*` and enabling a shorthand version with the `:` prefix.

Over the last 5 years, we have learned how LWC developers are using the existing directives. And as part of the harmonization process, I would like to make sure we also fix some of the design flaws in the `for:each` directive:
- Have the `lwc:key` directive applied to the same element as the `lwc:each` directive. Currently, the `key` is applied to the children elements, which makes very little sense and prevents some optimizations like the usage of fragments in the diffing algo.
- Make the `key` directive optional. I have seen components having to create a unique id using `Math.random()` just to use it as a value for the `key` directive.
- Exposing the current index of the iteration without having to use the `iterator:*` directive.

**Comment by AllanOricil** [comment]
> As always thanks for raising this @AllanOricil. 
> 
> I am also up for finishing the migration of all the LWC directives to `lwc:*` and enabling a shorthand version with the `:` prefix.
> 
> Over the last 5 years, we have learned how LWC developers are using the existing directives. And as part of the harmonization process, I would like to make sure we also fix some of the design flaws in the `for:each` directive:
> - Have the `lwc:key` directive applied to the same element as the `lwc:each` directive. Currently, the `key` is applied to the children elements, which makes very little sense and prevents some optimizations like the usage of fragments in the diffing algo.
> - Make the `key` directive optional. I have seen components having to create a unique id using `Math.random()` just to use it as a value for the `key` directive.
> - Exposing the current index of the iteration without having to use the `iterator:*` directive.

Another question, If I mutate an array that is used in a `for:each` does lwc recreate all the elements in the DOM, or does it patch and remove the ones that are no longer used?
For example: a for loop rendered 1000 todo items, and then the todo Items array was mutaded

 ````js
this.todoItems = this.todoItems.filter((t)=>t.include))
````

Does lwc throw away all 1000 elements from the DOM and then mount new ones, or does it patch some elements and unmount the ones that were deleted from the array? 

Not sure how Vue does it, but their doc says it has a "smart heurist algorithm" that it is used to avoid unnecessary changes to the DOM when an array used in a for loop is mutated. So, maybe this could be another good optimization you could implement.

**Comment by pmdartus** [comment]
> Another question, If I mutate an array that is used in a `for:each` does lwc recreate all the elements in the DOM, or does it patch and remove the ones that are no longer used?

Yes, it's something that is baked into LWC rendering engine from day 1. 
This is the reason why template iterations require a stable `key` directive. This key enables the engine to map items between 2 rendering cycles to apply the minimal set of DOM operations. 

You can find this logic in [here in LWC](https://github.com/salesforce/lwc/blob/8f3d5c95cc3438f3482f816c2b9ce2ffdaea3050/packages/%40lwc/engine-core/src/framework/rendering.ts#L837-L942), and [here in Vue](https://github.com/vuejs/core/blob/a1b3057c4deeb89b371ae418e2da557629d10d88/packages/runtime-core/src/renderer.ts#L1751).
